sdk/core/System.ClientModel/src/Convenience/ClientCache.cs

```diff
@@ -33,15 +30,13 @@ public ClientCache(int maxSize = 100)
     /// Updates the last-used timestamp on every hit.
     /// </summary>
     /// <typeparam name="T">The type of the client.</typeparam>
+    /// <param name="clientId">An equality-comparable key representing the client configuration.</param>
     /// <param name="createClient">A factory function to create the client if not cached.</param>
-    /// <param name="id">An identifier for the client instance.</param>
     /// <returns>The cached or newly created client instance.</returns>
-    public T GetClient<T>(Func<T> createClient, string? id) where T : class
+    public T GetClient<T>(IEquatable<object> clientId, Func<T> createClient) where T : class
     {
-        (Type, string) key = (typeof(T), id ?? string.Empty);
-
         // If the client exists, update its timestamp.
-        if (_clients.TryGetValue(key, out var cached))
+        if (_clients.TryGetValue(clientId, out var cached))
         {
             cached.LastUsed = Stopwatch.GetTimestamp();
             return (T)cached.Client;
```

Comment:
instead of just casting, can you check the type of the cached lient and if it's not T, throw an InvalidOperationException saying that the clientId is associated with client of type cached.Client.GetType and not T